### PR TITLE
Accept quoted int or float values as strings

### PR DESCRIPTION
**Pull request by ivanpauno**
Fixes https://github.com/ros2/rcl/issues/461.
Replaces https://github.com/ros2/rcl/pull/251.

This is #251 rebased with master, and with the bug fix over it.

Similar to https://github.com/ros2/rcl/pull/415.

### CODE REVIEW

**Comment by ivanpauno** [comment]
CI, just to keep the things moving (up to rcl, only fastrtps):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7434/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3579/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6096/)

File: rcl_yaml_param_parser/test/test_parse_yaml.cpp [hunk 1]
```diff
@@ -43,6 +43,24 @@ TEST(test_file_parser, correct_syntax) {
   allocator.deallocate(path, allocator.state);
 }
 
+TEST(test_file_parser, string_array_with_quoted_number) {
+  rcutils_reset_error();
+  EXPECT_TRUE(rcutils_get_cwd(cur_dir, 1024));
+  char * test_path = rcutils_join_path(cur_dir, "test");
```
Comment (jacobperron) [inline rcl_yaml_param_parser/test/test_parse_yaml.cpp]: Doesn't [`rcutils_join_path`](https://github.com/ros2/rcutils/blob/7350a3b3215788899f046fa201081db52f003a92/include/rcutils/filesystem.h#L124) need an allocator as a third parameter? :thinking:

File: rcl_yaml_param_parser/test/test_parse_yaml.cpp [hunk 2]
```diff
@@ -43,6 +43,24 @@ TEST(test_file_parser, correct_syntax) {
...
+  char * path = rcutils_join_path(test_path, "string_array_with_quoted_number.yaml");
+  fprintf(stderr, "cur_path: %s\n", path);
+  EXPECT_TRUE(rcutils_exists(path));
+  rcl_params_t * params_hdl = rcl_yaml_node_struct_init(allocator);
+  ASSERT_TRUE(params_hdl);
```
Comment (jacobperron) [inline rcl_yaml_param_parser/test/test_parse_yaml.cpp]: EXPECT_TRUE would be better so we free resources below.

File: rcl_yaml_param_parser/test/test_parse_yaml.cpp [hunk 3]
```diff
@@ -43,6 +43,24 @@ TEST(test_file_parser, correct_syntax) {
...
+  bool res = rcl_parse_yaml_file(path, params_hdl);
+  fprintf(stderr, "%s\n", rcutils_get_error_string_safe());
+  ASSERT_TRUE(res);
```
Comment (jacobperron) [inline rcl_yaml_param_parser/test/test_parse_yaml.cpp]: Same here

**Comment by ivanpauno** [comment]
Thank you @jacobperron for the review. I realized that I built/ran tests up to `rcl`, instead of to `rcl_yaml_param_parser` (both locally and in CI). I was actually testing nothing :joy: :man_facepalming:.

CI (up to rcl_yaml_param_parser, only fastrtps):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7438/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3583/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6100/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7274/)

File: rcl_yaml_param_parser/test/test_parse_yaml.cpp [hunk 4]
```diff
@@ -43,6 +43,26 @@ TEST(test_file_parser, correct_syntax) {
   allocator.deallocate(path, allocator.state);
 }
 
+TEST(test_file_parser, string_array_with_quoted_number) {
+  rcutils_reset_error();
+  EXPECT_TRUE(rcutils_get_cwd(cur_dir, 1024));
+  char * test_path = rcutils_join_path(cur_dir, "test", allocator);
```
Comment (jacobperron) [inline rcl_yaml_param_parser/test/test_parse_yaml.cpp]: nitpick: It appears all of the other tests are using local variables for the allocator. Maybe we can remove the global variable?

File: rcl_yaml_param_parser/test/test_parse_yaml.cpp [hunk 5]
```diff
@@ -43,6 +43,26 @@ TEST(test_file_parser, correct_syntax) {
...
+  EXPECT_TRUE(params_hdl);
+  if (params_hdl) {
+    bool res = rcl_parse_yaml_file(path, params_hdl);
+    fprintf(stderr, "%s\n", rcutils_get_error_string().str);
+    EXPECT_TRUE(res);
+    rcl_yaml_node_struct_print(params_hdl);
+    rcl_yaml_node_struct_fini(params_hdl);
+  }
+  free(test_path);
+  free(path);
```
Comment (jacobperron) [inline rcl_yaml_param_parser/test/test_parse_yaml.cpp]: Now that we're using an allocator, these two lines should be changed to use `allocator.deallocate()`.

**Comment by ivanpauno** [comment]
CI (up to rcl_yaml_param_parser, only fastrtps):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7451/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3596/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6114/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7289/)

**Pull request merged**